checks.d/hdfs_datanode.py

```diff
@@ -0,0 +1,169 @@
+
+# URL Paths
+JMX_PATH = 'jmx'
+
+# Metric types
+GAUGE = 'gauge'
+
+# HDFS bean name
+HDFS_DATANODE_BEAN_NAME = 'Hadoop:service=DataNode,name=FSDatasetState'
+
+# HDFS metrics
+HDFS_METRICS = {
+    'Remaining' : ('hdfs_datanode.dfs_remaining',  GAUGE),
+    'Capacity' :('hdfs_datanode.dfs_capacity', GAUGE),
+    'DfsUsed' : ('hdfs_datanode.dfs_used', GAUGE),
+    'CacheCapacity' : ('hdfs_datanode.cache_capacity', GAUGE),
+    'CacheUsed' : ('hdfs_datanode.cache_used', GAUGE),
+    'NumFailedVolumes' : ('hdfs_datanode.num_failed_volumes', GAUGE),
+    'LastVolumeFailureDate' : ('hdfs_datanode.last_volume_failure_date', GAUGE),
+    'EstimatedCapacityLostTotal' : ('hdfs_datanode.estimated_capacity_lost_total', GAUGE),
+    'NumBlocksCached' : ('hdfs_datanode.num_blocks_cached', GAUGE),
+    'NumBlocksFailedToCache' : ('hdfs_datanode.num_blocks_failed_to_cache', GAUGE),
+    'NumBlocksFailedToUnCache' : ('hdfs_datanode.num_blocks_failed_to_uncache', GAUGE)
+}
+
+class HDFSDataNode(AgentCheck):
+
+    def check(self, instance):
+        jmx_address = instance.get('hdfs_datanode_jmx_uri')
+        if jmx_address is None:
+            raise Exception('The JMX URL must be specified in the instance configuration')
+
+        # Get metrics from JMX
+        self._hdfs_datanode_metrics(jmx_address)
+
+    def _hdfs_datanode_metrics(self, jmx_uri):
+        '''
+        Get HDFS data node metrics from JMX
+        '''
+        response = self._rest_request_to_json(jmx_uri,
+            JMX_PATH,
+            query_params={'qry':'Hadoop:service=DataNode,name=FSDatasetState'})
+
+        beans = response.get('beans', [])
+
+        tags = ['datanode_url:' + jmx_uri]
+
+        for bean in beans:
+            bean_name = bean.get('name')
+
+            if bean_name == HDFS_DATANODE_BEAN_NAME:
+                for metric, (metric_name, metric_type) in HDFS_METRICS.iteritems():
+                    metric_value = bean.get(metric)
+
+                    if metric_value is not None:
+                        self._set_metric(metric_name, metric_type, metric_value, tags)
+
+    def _set_metric(self, metric_name, metric_type, value, tags=None, device_name=None):
+        '''
+        Set a metric
+        '''
+        if metric_type == GAUGE:
+            self.gauge(metric_name, value, tags=tags, device_name=device_name)
+        else:
+            self.log.error('Metric type "%s" unknown' % (metric_type))
+
+    def _rest_request_to_json(self, address, object_path, query_params):
+        '''
+        Query the given URL and return the JSON response
+        '''
+        response_json = None
+
+        service_check_tags = ['instance:%s' % self.hostname]
+
+        url = address
+
+        if object_path:
+            url = self._join_url_dir(url, object_path)
+
+        # Add query_params as arguments
+        if query_params:
+            query = '&'.join(['{}={}'.format(key, value) for key, value in query_params.iteritems()])
```

Comment:
We support python 2.6 so you have to use `'{0}={1}'.format(key, value)` here (i.e. number the fields)